lib/util/entrypoints.js

```diff
@@ -0,0 +1,605 @@
+	}
+	if (request.charCodeAt(request.length - 1) === slashCode) {
+		throw new Error("Only requesting file allowed");
+	}
+}
+
+/**
+ * @param {string} request request
+ */
+function assertImportsFieldRequest(request) {
+	if (request.charCodeAt(0) !== hashCode) {
+		throw new Error('Request should start with "#"');
+	}
+	if (request.length === 1) {
+		throw new Error("Request should have at least 2 characters");
+	}
+	if (request.charCodeAt(1) === slashCode) {
+		throw new Error('Request should not start with "#/"');
+	}
+	if (request.charCodeAt(request.length - 1) === slashCode) {
+		throw new Error("Only requesting file allowed");
+	}
+}
+
+/**
+ * @param {string} exp export target
+ * @param {boolean} expectFolder is folder expected
+ */
+function assertExportTarget(exp, expectFolder) {
+	if (
+		exp.charCodeAt(0) === slashCode ||
+		(exp.charCodeAt(0) === dotCode && exp.charCodeAt(1) !== slashCode)
+	) {
+		throw new Error(
+			`Export should be relative path and start with "./", got ${JSON.stringify(
+				exp
+			)}.`
+		);
+	}
+
+	const isFolder = exp.charCodeAt(exp.length - 1) === slashCode;
+
+	if (isFolder !== expectFolder) {
+		throw new Error(
+			expectFolder
+				? `Expecting folder to folder mapping. ${JSON.stringify(
+						exp
+				  )} should end with "/"`
+				: `Expecting file to file mapping. ${JSON.stringify(
+						exp
+				  )} should not end with "/"`
+		);
+	}
+}
+
+/**
+ * @param {string} imp import target
+ * @param {boolean} expectFolder is folder expected
+ */
+function assertImportTarget(imp, expectFolder) {
+	const isFolder = imp.charCodeAt(imp.length - 1) === slashCode;
+
+	if (isFolder !== expectFolder) {
+		throw new Error(
+			expectFolder
+				? `Expecting folder to folder mapping. ${JSON.stringify(
+						imp
+				  )} should end with "/"`
+				: `Expecting file to file mapping. ${JSON.stringify(
+						imp
+				  )} should not end with "/"`
+		);
+	}
+}
+
+/**
+ * Trying to match request to field
+ * @param {string} request request
+ * @param {PathTreeNode} treeRoot path tree root
+ * @returns {[MappingValue, number]|null} match or null
+ */
+function findMatch(request, treeRoot) {
+	if (request.length === 1) {
+		const value = treeRoot.files.get("*root*");
+
+		return value ? [value, 1] : null;
+	}
+
+	if (treeRoot.children === null && treeRoot.folder === null) {
+		const value = treeRoot.files.get(request);
+
+		return value ? [value, request.length] : null;
+	}
+
+	let node = treeRoot;
+	let lastNonSlashIndex = 0;
+	let slashIndex = request.indexOf("/", 2);
+
+	/** @type {[MappingValue, number]|null} */
+	let lastFolderMatch = null;
+
+	while (slashIndex !== -1) {
+		const folder = request.slice(lastNonSlashIndex, slashIndex);
+
+		const folderMapping = node.folder;
+		if (folderMapping) {
+			if (lastFolderMatch) {
+				lastFolderMatch[0] = folderMapping;
+				lastFolderMatch[1] = lastNonSlashIndex;
+			} else {
+				lastFolderMatch = [folderMapping, lastNonSlashIndex || 2];
```

Comment:
this is needed to concatenate without "./". This matching could be implemented other way with request "normalization" (without "./"), but this will require slicing request string and some extra logic too. So I think this design is fine enough and performs faster (less allocations)